pkg/provision/cleanup.go

```diff
@@ -0,0 +1,234 @@
+			if !ok {
+				log.Info().Msgf("skipping volume '%s' has no assigned quota", subvol.Path)
+				continue
+			}
+
+			// Now, is this subvol in one of the toSave ?
+			if _, ok := toSave[filepath.Base(subvol.Path)]; ok {
+				log.Info().Msgf("skipping volume '%s' is used", subvol.Path)
+				continue
+			}
+
+			// Now, is this subvol in one of the toDelete ?
+			if _, ok := toDelete[filepath.Base(subvol.Path)]; ok {
+				// delete the subvolume
+				delete := checkReservationToDelete(subvol.Path, client)
+				if delete {
+					log.Info().Msgf("deleting subvolume %s", subvol.Path)
+					if err := utils.SubvolumeRemove(ctx, filepath.Join(path, subvol.Path)); err != nil {
+						log.Err(err).Msgf("failed to delete subvol '%s'", subvol.Path)
+					}
+					if err := utils.QGroupDestroy(ctx, qgroup.ID, path); err != nil {
+						log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+					}
+					continue
+				}
+				log.Info().Msgf("skipping subvolume %s", subvol.Path)
+				continue
+			}
+
+			// Is this subvol not in toDelete and not in toSave?
+			// Check the explorer if it needs to be deleted
+			delete := checkReservationToDelete(subvol.Path, client)
+			if delete {
+				log.Info().Msgf("deleting subvolume %s", subvol.Path)
+				if err := utils.SubvolumeRemove(ctx, filepath.Join(path, subvol.Path)); err != nil {
+					log.Err(err).Msgf("failed to delete subvol '%s'", subvol.Path)
+				}
+				if err := utils.QGroupDestroy(ctx, qgroup.ID, path); err != nil {
+					log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+				}
+				continue
+			}
+			log.Info().Msgf("skipping subvolume %s", subvol.Path)
+		}
+	}
+
+	return nil
+}
+
+func checkReservationToDelete(path string, cl *client.Client) bool {
+	log.Info().Msgf("checking explorer for reservation: %s", path)
+	reservation, err := cl.Workloads.NodeWorkloadGet(path)
+	if err != nil {
+		var hErr client.HTTPError
+		if ok := errors.As(err, &hErr); ok {
+			resp := hErr.Response()
+			// If reservation is not found it should be deleted
+			if resp.StatusCode == 404 {
+				return true
+			}
+		}
+		return false
+	}
+
+	if reservation.GetNextAction() == workloads.NextActionDelete {
+		log.Info().Msgf("subvolume with path %s has next action to delete", path)
+		return true
+	}
+
+	return false
+}
+
+// checks running containers for subvolumes that might need to be saved because they are used
+// and subvolumes that might need to be deleted because they have no attached container anymore
+func checkContainers(msgBrokerCon string) (map[string]struct{}, map[string]struct{}, error) {
+	toSave := make(map[string]struct{})
+	toDelete := make(map[string]struct{})
+	zbus, err := zbus.NewRedisClient(msgBrokerCon)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	client, err := containerd.New(containerdSock)
+	if err != nil {
+		log.Err(err).Msgf("failed to create containerd connection")
+		return nil, nil, err
+	}
+
+	ns, err := client.NamespaceService().List(context.Background())
```

Comment:
create one context `ctx := context.Background()` and then pass it to all methods that needs it. Or even better, change the function signature `checkContainers` to accept a context object.